cpp/src/io/parquet/reader_impl_chunking.cu

```diff
@@ -341,14 +406,963 @@ template <typename T = uint8_t>
+ * shot because we do not know how many rows we actually have in the pages of list columns.
+ * So we have to make an educated guess that fits within the memory limits, and then adjust
+ * for subsequent subpasses when we see how many rows we actually receive.
+ *
+ * @param c_info The cumulative page size information (row count and byte size) per column
+ * @param pages All of the pages in the pass
+ * @param page_offsets Offsets into the pages array representing the first page for each column
+ * @param start_row The row to start the subpass at
+ * @param size_limit The size limit in bytes of the subpass
+ * @param num_columns The number of columns
+ * @param stream The stream to execute cuda operations on
+ * @returns A tuple containing a vector of page_span structs indicating the page indices to include
+ * for each column to be processed, the total number of pages over all columns, and the total
+ * expected memory usage (including scratch space)
+ *
+ */
+std::tuple<std::vector<page_span>, size_t, size_t> compute_next_subpass(
+  rmm::device_uvector<cumulative_page_info> const& c_info,
+  cudf::detail::hostdevice_vector<PageInfo> const& pages,
+  cudf::detail::hostdevice_vector<size_type> const& page_offsets,
+  size_t start_row,
+  size_t size_limit,
+  size_t num_columns,
+  rmm::cuda_stream_view stream)
+{
+  auto [aggregated_info, page_keys_by_split] = adjust_cumulative_sizes(c_info, pages, stream);
+
+  // bring back to the cpu
+  auto const h_aggregated_info = cudf::detail::make_std_vector_sync(aggregated_info, stream);
+  // print_cumulative_row_info(h_aggregated_info, "adjusted");
+
+  // TODO: if the user has explicitly specified skip_rows/num_rows we could be more intelligent
+  // about skipping subpasses/pages that do not fall within the range of values, but only if the
+  // data does not contain lists (because our row counts are only estimates in that case)
+
+  // find the next split
+  auto const start_index = find_start_index(h_aggregated_info, start_row);
+  auto const cumulative_size =
+    start_row == 0 || start_index == 0 ? 0 : h_aggregated_info[start_index - 1].size_bytes;
+  auto const end_index =
+    find_next_split(start_index, start_row, cumulative_size, h_aggregated_info, size_limit);
+  auto const end_row = h_aggregated_info[end_index].row_index;
+
+  // for each column, collect the set of pages that spans start_row / end_row
+  rmm::device_uvector<page_span> page_bounds(num_columns, stream);
+  auto iter = thrust::make_counting_iterator(size_t{0});
+  auto page_row_index =
+    cudf::detail::make_counting_transform_iterator(0, get_page_row_index{c_info});
+  thrust::transform(rmm::exec_policy_nosync(stream),
+                    iter,
+                    iter + num_columns,
+                    page_bounds.begin(),
+                    get_page_span{page_offsets, page_row_index, start_row, end_row});
+  auto h_page_bounds = cudf::detail::make_std_vector_sync(page_bounds, stream);
```

Comment:
So we transfer to the host for further processing. Can we do further processing on device?

Reply:
Actually, I confused this with the other comment https://github.com/rapidsai/cudf/pull/14360#discussion_r1464087196

The output from here is a series of spans used to copy pages around using std::copy().   This could in theory be done on the gpu, but it's a multi-buffer copy so it would require a little bit of thought.  I think this would be better suited as a followup.